### PR TITLE
Support domain deformations in elliptic solver

### PR DESCRIPTION
**Pull request by nilsvu**
## Proposed changes

This enables horizon-conforming (ellipsoidal) excision surfaces for spinning BH initial data. It will also allow for surface-fitting coordinates for BNS initial data. For reference, here's my thesis branch that had this implemented in a different way (using static deformed maps, not time dependence): https://github.com/sxs-collaboration/spectre/compare/develop...nilsvu:spectre:thesis

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: src/Domain/ElementMap.cpp [hunk 1]
```diff
@@ -53,6 +55,52 @@ ElementMap<Dim, TargetFrame>::ElementMap(
...
+template <size_t Dim, typename TargetFrame>
+ElementMap<Dim, TargetFrame>::ElementMap(ElementId<Dim> element_id,
+                                         const Block<Dim>& block)
+    : ElementMap(
+          std::move(element_id),
+          [&block]() -> std::unique_ptr<domain::CoordinateMapBase<
+                         Frame::BlockLogical, TargetFrame, Dim>> {
+            if constexpr (std::is_same_v<TargetFrame, Frame::Inertial>) {
+              if (block.is_time_dependent()) {
+                if (block.has_distorted_frame()) {
```
Comment (nilsdeppe) [inline src/Domain/ElementMap.cpp]: Do you actually need this case? Shouldn't the else branch work even if there's a distorted frame? I'm fairly confident the evolution code at least requires grid->inertial to always be a valid map, independent of whether or not there's a distorted frame.

File: src/Domain/CMakeLists.txt [hunk 2]
```diff
@@ -70,15 +70,14 @@ target_link_libraries(
   DomainBoundaryConditions
```
Comment (nilsdeppe) [inline src/Domain/CMakeLists.txt:70]: In the elliptic code do you actually evaluate the maps outside of compute items? If not, this commit, if not the previous as well could be replaced by adding a grid->inertial map into the elliptic box and using the time-dependent compute tags we use in the evolution code. Those tags don't do any extra work for "static" domains. I guess they currently require a `Tags::Time` to be in the box, but we could reasonably template on a bool or something to elide the time if it's not needed.

File: src/Elliptic/DiscontinuousGalerkin/Initialization.cpp [hunk 3]
```diff
@@ -50,29 +50,26 @@ void InitializeGeometry<Dim>::operator()(
...
   *element = domain::Initialization::create_initial_element(element_id, block,
                                                             initial_refinement);
   // Element map
-  *element_map = ElementMap<Dim, Frame::Inertial>{
-      element_id, block.stationary_map().get_clone()};
+  *element_map = ElementMap<Dim, Frame::Inertial>{element_id, block};
   // Coordinates
   *logical_coords = logical_coordinates(*mesh);
-  *inertial_coords = element_map->operator()(*logical_coords);
+  *inertial_coords =
```
Comment (nilsdeppe) [inline src/Elliptic/DiscontinuousGalerkin/Initialization.cpp:68]: Right, if you don't need the map in other places, this seems like the perfect place to just call the grid->inertial map once

**Review by nilsdeppe — CHANGES_REQUESTED** [review]
Looks reasonable, I do have one design question. Might be easier to discuss briefly on a call, up to you. Squash any changes directly, please.

File: src/Domain/ElementMap.cpp (hunk 1 above)
Comment (nilsvu) [reply src/Domain/ElementMap.cpp]: You're right 👍 removed the unnecessary case.

File: src/Domain/CMakeLists.txt (hunk 2 above)
Comment (nilsvu) [reply src/Domain/CMakeLists.txt:70]: Currently the map is only needed during initialization; in multiple actions though (domain init, subdomain init, faces & mortars). I thought about removing the ElementMap tag from the DataBox, but kept it for now so I don't have to recreate the ElementMap in these places. Could be done though I think. When we update the map parameters we have to re-evaluate the map as well though, which might be easier if the ElementMap is available. I haven't thought that part through yet.

Comment (nilsdeppe) [reply src/Domain/CMakeLists.txt:70]: Alright, that sounds good. Thanks for the explanation!

**Review by nilsdeppe — DISMISSED** [review]

**Comment by nilsvu** [comment]
Fixed a use-after-move clang-tidy complaint in the ElementMap constructor `ASSERT`. Nice to have clang tidy to catch these things.

**Review by nilsdeppe — APPROVED** [review]

**Pull request merged**